### ISSUE TITLE
Refactor rollbacks to apply to inputs at the stack or game level, not garbage

### ISSUE THREAD

**Issue by sharpobject** [panel-attack/panel-game#29]
Sending garbage into the past is confusing. Let's delete it.

Instead, send inputs into the past. See these resources for how this should work: http://mauve.mizuumi.net/2012/07/05/understanding-fighting-game-networking https://www.youtube.com/watch?v=7jb0FOcImdg https://www.youtube.com/watch?v=k9JTIn1SVQ4

This lets us keep only one old copy of the gamestate around, rather than a pool of copies, since the "next input" can only possibly go to one state.

**Comment by JamesVanBoxtel** [comment]
I had a version of this working for telegraph, but it was just too slow to rollback inputs all the time. Given we only need to rollback when attacks happen, I'm not sure its the right approach after playing with it.

**Comment by Endaris** [comment]
I sat through the premise of the GDC video and I think this approach is definitely inferior for Panel Attack compared to the current one.
Inputs in a fighting game are time critical, whatever input a player does, typically telegraphs a motion that gives critical information to the other player, making it the core for interactive gameplay.
PA gameplay has intrinsically low interactivity for the majority of the game:
The most relevant telegraphs in PA are the fanfares and combo/chain SFX alongside the pop SFX.
The game should thus work on delivering these as fast as possible at any point in time.
With how the game currently handles running delayed incoming inputs, we can be informed of a sound telegraph up to 21 frames late before the game is aborted for desync. A solution to speed this up would be to up the max amount of runs per frame while catching up with delayed inputs coming in from the network.

Ultimately, even if the solution was implemented, garbage would effectively *still* be sent to the past. The only way around that is to lower the maximum allowed latency between player before a desync occurs (in the presentation example, 10 frames) which is very harsh considering we aren't using P2P but a server approach, increasing our general expected delay to at least twice that.

Another potential issue with rolling back another persons stack by a greater amount of frames would be that you might actually not receive older telegraphs at all or at the same time which brings its own set of problems.

I measured a relatively short game of pusan vs Malfurner I spectated the other day where at game end 72 seconds in, pusan sits at 194 panels popped with 32.3 gpm and 552 apm. I think it is valid to consider that peak level. On average that makes less than 3 panels destroyed in a second which supports the assumption that the visual change in the stack is expected to be relatively unimportant for the sake of screenwatching purpose, making the clean playback of SFX telegraphs the higher priority.

I'm not saying this is impossible to implement but we'd need a major overhaul of the engine to ensure that rollback is performant by facilitating state checks and ideally a P2P connection to ensure that latency can be low enough for things to be actually worth it and even then we might only want to rollback inputs for latency spikes upwards of the garbage drop rollback threshold.

If there is one thing we might want to look out for changing the current behaviour is covering cases better where a garbage rollback is potentially avoidable by catching up with the delayed inputs faster, e.g. creating a copy of the stack and `simulate` running it in fast forward to check if garbage would drop during the catchup process and if that is the case, ramp up `max_runs_per_frame` to a degree where the garbage can drop normally - resulting distorted SFX telegraphs would still be preferable over getting your stack rolled back for the player so that might be worth looking into but that is likely still something for a different issue.

Panel Attack is not a fighting game and with this I'm going to close this.
